File: tool/generate_icon.dart
import 'dart:io';
import 'package:image/image.dart';

void main() {
  _generate('assets/launcher_icon/icon.png', foreground: false);
  _generate('assets/launcher_icon/icon_foreground.png', foreground: true);
  print('Done: assets/launcher_icon/icon.png + icon_foreground.png');
}

void _generate(String path, {required bool foreground}) {
  const size = 1024;
  final img = Image(width: size, height: size);

  // Background
  if (foreground) {
    fill(img, color: ColorRgba8(0, 0, 0, 0)); // transparent for adaptive
  } else {
    fill(img, color: ColorRgb8(10, 10, 26));
    _backgroundGlow(img, size);
  }

  const cx = size ~/ 2;
  const cy = size ~/ 2;

  // Gap / bar layout
  const gapY = cy + 30;
  const barH = 75;
  const gapHalf = 105;
  const barLeft = cx - gapHalf;
  const barRight = cx + gapHalf;

  // Bar glow layers
  for (var g = 24; g >= 2; g -= 2) {
    final alpha = (8 + (24 - g)).clamp(0, 255);
    fillRect(img,
        x1: 0,
        y1: gapY - barH ~/ 2 - g,
        x2: barLeft + g ~/ 2,
        y2: gapY + barH ~/ 2 + g,
        color: ColorRgba8(0, 200, 255, alpha));
    fillRect(img,
        x1: barRight - g ~/ 2,
        y1: gapY - barH ~/ 2 - g,
        x2: size,
        y2: gapY + barH ~/ 2 + g,
        color: ColorRgba8(0, 200, 255, alpha));
  }

  // Bars (solid neon blue)
  fillRect(img,
      x1: 0,
      y1: gapY - barH ~/ 2,
      x2: barLeft,
      y2: gapY + barH ~/ 2,
      color: ColorRgb8(0, 190, 245));
  fillRect(img,
      x1: barRight,
      y1: gapY - barH ~/ 2,
      x2: size,
      y2: gapY + barH ~/ 2,
      color: ColorRgb8(0, 190, 245));

  // Bar edge highlight (top edge brighter)
  fillRect(img,
      x1: 0, y1: gapY - barH ~/ 2, x2: barLeft, y2: gapY - barH ~/ 2 + 4,
      color: ColorRgb8(80, 230, 255));
  fillRect(img,
      x1: barRight, y1: gapY - barH ~/ 2, x2: size, y2: gapY - barH ~/ 2 + 4,
      color: ColorRgb8(80, 230, 255));

  // Ball glow layers
  const ballR = 58;
  for (var g = 48; g >= 4; g -= 4) {
    final alpha = ((1 - g / 52) * 55).toInt().clamp(0, 255);
    fillCircle(img,
        x: cx,
        y: gapY,
        radius: ballR + g,
        color: ColorRgba8(180, 235, 255, alpha));
  }

  // Ball inner glow
  fillCircle(img, x: cx, y: gapY, radius: ballR + 8,
      color: ColorRgba8(220, 245, 255, 120));

  // Ball (white)
  fillCircle(img, x: cx, y: gapY, radius: ballR,
      color: ColorRgb8(255, 255, 255));

  // Ball highlight (top-left bright spot)
  fillCircle(img, x: cx - 18, y: gapY - 18, radius: 18,
      color: ColorRgba8(255, 255, 255, 160));

  // Trail dots above ball
  const trailColor = ColorRgba8(180, 235, 255, 140);
  for (var i = 1; i <= 5; i++) {
    final tr = (ballR * (1 - i * 0.16)).toInt().clamp(4, ballR);
    final alpha = (140 - i * 22).clamp(20, 140);
    fillCircle(img,
        x: cx,
        y: gapY - ballR * 2 - i * (ballR ~/ 2 + 10),
        radius: tr,
        color: ColorRgba8(180, 235, 255, alpha));
  }

  final dir = Directory('assets/launcher_icon');
  if (!dir.existsSync()) dir.createSync(recursive: true);
  File(path).writeAsBytesSync(encodePng(img));
}

void _backgroundGlow(Image img, int size) {
  final cx = size ~/ 2;
  final cy = size ~/ 2;
  for (var r = 420; r > 20; r -= 20) {
    final t = 1 - r / 420;
    final alpha = (t * 28).toInt().clamp(0, 255);
    fillCircle(img, x: cx, y: cy, radius: r,
        color: ColorRgba8(20, 15, 55, alpha));
  }
}

File: tool/generate_icon.py
from PIL import Image, ImageDraw, ImageFilter, ImageFont
import os, math, random

SIZE = 1024
os.makedirs("assets/launcher_icon", exist_ok=True)

NEON_BLUE  = (0, 200, 255)
NEON_PINK  = (255, 45, 135)
WHITE      = (255, 255, 255)
BG         = (10, 10, 26)

def add_layer(base: Image.Image, layer: Image.Image) -> Image.Image:
    return Image.alpha_composite(base, layer)

def glow_rect(draw, x1, y1, x2, y2, color, spread=28):
    for g in range(spread, 0, -3):
        a = int((1 - g / (spread + 2)) * 80)
        draw.rectangle([x1-g, y1-g, x2+g, y2+g], fill=(*color, a))

def glow_circle(draw, cx, cy, r, color, spread=50, max_alpha=100):
    for g in range(spread, 0, -4):
        a = int((1 - g / (spread + 2)) * max_alpha)
        draw.ellipse([cx-r-g, cy-r-g, cx+r+g, cy+r+g], fill=(*color, a))

def make_icon(transparent_bg: bool) -> Image.Image:
    img = Image.new("RGBA", (SIZE, SIZE), (0, 0, 0, 0))

    # ── Background ──────────────────────────────────────────────
    if not transparent_bg:
        bg = Image.new("RGBA", (SIZE, SIZE), (*BG, 255))

        # Stars
        rng = random.Random(42)
        sd = ImageDraw.Draw(bg)
        for _ in range(180):
            x = rng.randint(0, SIZE)
            y = rng.randint(0, SIZE)
            r = rng.choice([1, 1, 1, 2])
            a = rng.randint(60, 200)
            sd.ellipse([x-r, y-r, x+r, y+r], fill=(255, 255, 255, a))

        # Radial background glow (center deep purple)
        for radius in range(480, 0, -20):
            t = 1 - radius / 480
            a = int(t * 38)
            sd.ellipse([SIZE//2-radius, SIZE//2-radius,
                        SIZE//2+radius, SIZE//2+radius],
                       fill=(28, 10, 70, a))

        img = add_layer(img, bg)

    cx = SIZE // 2

    # ── Three bars at different Y positions ─────────────────────
    bars = [
        # (y_center, gap_half, color, alpha_mult)
        (SIZE//2 - 220, 115,  NEON_PINK,  0.45),  # top bar, pink, faded
        (SIZE//2 + 10,  108,  NEON_BLUE,  1.0),   # middle bar, blue, hero
        (SIZE//2 + 230, 130,  NEON_BLUE,  0.35),  # bottom bar, faded
    ]

    for (bar_y, gap_half, color, am) in bars:
        bar_h  = 68
        bar_l  = cx - gap_half
        bar_r  = cx + gap_half

        # Glow
        glow = Image.new("RGBA", (SIZE, SIZE), (0, 0, 0, 0))
        gd   = ImageDraw.Draw(glow)
        glow_rect(gd, 0, bar_y-bar_h//2, bar_l, bar_y+bar_h//2, color, spread=30)
        glow_rect(gd, bar_r, bar_y-bar_h//2, SIZE, bar_y+bar_h//2, color, spread=30)
        glow = glow.filter(ImageFilter.GaussianBlur(8))
        # Apply alpha multiplier
        r2, g2, b2, a2 = glow.split()
        a2 = a2.point(lambda x: int(x * am))
        glow = Image.merge("RGBA", (r2, g2, b2, a2))
        img = add_layer(img, glow)

        # Solid bars
        solid = Image.new("RGBA", (SIZE, SIZE), (0, 0, 0, 0))
        sd2   = ImageDraw.Draw(solid)
        bar_alpha = int(255 * am)
        sd2.rectangle([0, bar_y-bar_h//2, bar_l, bar_y+bar_h//2],
                      fill=(*color, bar_alpha))
        sd2.rectangle([bar_r, bar_y-bar_h//2, SIZE, bar_y+bar_h//2],
                      fill=(*color, bar_alpha))
        # Bright top edge
        edge_a = int(255 * am)
        sd2.rectangle([0, bar_y-bar_h//2, bar_l, bar_y-bar_h//2+5],
                      fill=(min(color[0]+80,255), min(color[1]+40,255), 255, edge_a))
        sd2.rectangle([bar_r, bar_y-bar_h//2, SIZE, bar_y-bar_h//2+5],
                      fill=(min(color[0]+80,255), min(color[1]+40,255), 255, edge_a))
        img = add_layer(img, solid)

    # ── Ball (centered in middle bar gap) ───────────────────────
    hero_y   = SIZE//2 + 10
    ball_r   = 62

    # Ball glow (blue only — white glow was overblowing bars)
    for glow_color, spread, alpha in [
        (NEON_BLUE, 55, 80),
        (WHITE,     22, 35),
    ]:
        gl = Image.new("RGBA", (SIZE, SIZE), (0, 0, 0, 0))
        gd = ImageDraw.Draw(gl)
        glow_circle(gd, cx, hero_y, ball_r, glow_color, spread=spread, max_alpha=alpha)
        gl = gl.filter(ImageFilter.GaussianBlur(7))
        img = add_layer(img, gl)

    ball_layer = Image.new("RGBA", (SIZE, SIZE), (0, 0, 0, 0))
    bd = ImageDraw.Draw(ball_layer)
    bd.ellipse([cx-ball_r, hero_y-ball_r, cx+ball_r, hero_y+ball_r],
               fill=(255, 255, 255, 255))
    # Highlight
    bd.ellipse([cx-30, hero_y-30-14, cx+6, hero_y+6-14],
               fill=(255, 255, 255, 170))
    img = add_layer(img, ball_layer)

    # Trail (below ball, neon blue fading)
    trail = Image.new("RGBA", (SIZE, SIZE), (0, 0, 0, 0))
    td = ImageDraw.Draw(trail)
    for i in range(1, 6):
        tr = max(5, int(ball_r * (1 - i * 0.16)))
        a  = max(18, 145 - i * 28)
        ty = hero_y + ball_r + 10 + i * (ball_r // 2 + 6)
        td.ellipse([cx-tr, ty-tr, cx+tr, ty+tr], fill=(*NEON_BLUE, a))
    img = add_layer(img, trail)

    # ── PULSE text ───────────────────────────────────────────────
    if not transparent_bg:
        font_path = "C:/Windows/Fonts/arialbd.ttf"
        try:
            font = ImageFont.truetype(font_path, size=130)
            tl = Image.new("RGBA", (SIZE, SIZE), (0, 0, 0, 0))
            td2 = ImageDraw.Draw(tl)
            text = "PULSE"
            bbox = td2.textbbox((0, 0), text, font=font)
            tw = bbox[2] - bbox[0]
            tx = (SIZE - tw) // 2 - bbox[0]
            ty_text = SIZE - 190

            # Text glow
            for spread in [16, 10, 5]:
                a = 60 if spread == 16 else (90 if spread == 10 else 130)
                td2.text((tx - spread//3, ty_text - spread//3),
                         text, font=font, fill=(*NEON_BLUE, a))
            # Text solid
            td2.text((tx, ty_text), text, font=font, fill=(255, 255, 255, 255))
            tl_blur = tl.filter(ImageFilter.GaussianBlur(3))
            img = add_layer(img, tl_blur)
            tl2 = Image.new("RGBA", (SIZE, SIZE), (0, 0, 0, 0))
            td3 = ImageDraw.Draw(tl2)
            td3.text((tx, ty_text), text, font=font, fill=(255, 255, 255, 255))
            img = add_layer(img, tl2)
        except Exception:
            pass  # skip text if font missing

    return img

full = make_icon(transparent_bg=False)
full.save("assets/launcher_icon/icon.png")
print("Saved: assets/launcher_icon/icon.png")

fg = make_icon(transparent_bg=True)
fg.save("assets/launcher_icon/icon_foreground.png")
print("Saved: assets/launcher_icon/icon_foreground.png")
